### ISSUE TITLE
dart2js: Eager initialization of static and top-level members

### ISSUE THREAD

**Issue by rakudrama** [dart-lang/sdk#35929]
dart2js directly initializes static variables that have a constant initializer, and uses lazy initialization for all other initializers. For example,

```dart
final List<dynamic> styles = ['...css...'];
```
compiles to
```js
    _lazy($, "styles64", "$get$styles1287", function() {
      return ["...css..."];
    });
```

The `styles` variable could be initialized eagerly since the expression always succeeds, makes no side effects, and the cost of evaluation at initial program load is small, probably _less_ that setting up the lazy evaluation.
```js
  $.styles64 = ["...css..."];
```

The optimization needs to make an early determination that an initialization can be eager since read accesses to lazy initialized static are calls rather than property loads, in this example, `$.$get$styles1287()` vs `$.styles64`. This could be quite limiting since many expressions become simple only through optimization.

It is difficult to anticipate if compiling an expression will generate a JavaScript expression. To avoid having a separate compiler for initializers, we can compile the expression like a lazy initializer and inspect the result. If the function contains a single return, that is the expression, otherwise we can call the function:
```js
$.styles64 = (function(){...;return ...})();
```
----
Historical note.

Lazy initializers for non-constants used to be necessary when dart2js supported multiple isolates running the same program: each isolate has its own variables and own (possibly mutable) values assigned to those variables. Even if that model was brought back, this optimization could still be applied for programs not using multiple isolates.

----
We can simplify some lazy initializers where we failed to anticipate the eager evaluation, but it turns out that the expression is a simple expression, and the variable is accessed only via the lazy initializer:
```js
    $.$get$styles1287 = alwaysReturns(["...css..."]);
```
We can tell the access patterns from the code generation impacts of the other methods.
Determining that the expression is simple is probably best done by analyzing the SSA just prior to codegen since that contains information about effects

**Comment by matanlurey** [comment]
+1.

We would also be happy to experiment with some sort of pragma that promises no side effects:

```dart
@pragma('dart2js:noSideEffects')
final List<dynamic> styles = ['...css...'];
```

**Comment by rakudrama** [comment]
@johnniwinther has implemented the first part of this: 068aa0d2e732f4fb6be72105cc727c82c448a55c.
It handles the Angular Dart CSS scenario quite well and reduces the size of one app I investigated by 0.29%
